drivers/include/drivers/DigitalOut.h

```diff
@@ -70,12 +79,27 @@ class DigitalOut {
         gpio_init_out_ex(&gpio, pin, value);
     }
 
+    /** Class destructor, deinitialize the pin
+     */
+#if MBED_CONF_DRIVERS_VIRTUALIZE_DIGITALOUT
```

Comment:
Rather than repeat this, arrange to have a MBED_DIGITALOUT_OPTIONAL_VIRTUAL and use that.

    #if MBED_CONF_VIRTUALIZE_DIGITALOUT
    #define MBED_DIGITALOUT_OPTIONAL_VIRTUAL virtual
    #else
    #define MBED_DIGITALOUT_OPTIONAL_VIRTUAL
    #endif

    MBED_DIGITALOUT_OPTIONAL_VIRTUAL
    ~DigitalOut()
    ...

Reply:
I know we've been working (at least on some customer branches) on the opposite problem - devirtualising virtual systems where we know we've got one specific implementation.

Not sure what there is in mainline master, but the conditional compilation patterns used during those optional devirtualisations - where you did start with the separate API and implementation classes but optionally merged them - would be something to look at.